README.md

# nanoDPO: Direct Preference Optimization for Time Series Data

[![PyPI](https://img.shields.io/pypi/v/nanoDPO.svg)](https://pypi.org/project/nanoDPO/)
[![Changelog](https://img.shields.io/github/v/release/jamesliu/nanoDPO?include_prereleases&label=changelog)](https://github.com/jamesliu/nanoDPO/releases)
[![Tests](https://github.com/jamesliu/nanoDPO/workflows/Test/badge.svg)](https://github.com/jamesliu/nanoDPO/actions?query=workflow%3ATest)
[![Documentation Status](https://readthedocs.org/projects/nanoDPO/badge/?version=stable)](http://nanoDPO.readthedocs.org/en/stable/?badge=stable)
[![License](https://img.shields.io/badge/license-Apache%202.0-blue.svg)](https://github.com/jamesliu/nanoDPO/blob/main/LICENSE)

## Introduction
Welcome to `nanoDPO` – a novel, cutting-edge library for Direct Preference Optimization (DPO) tailored for time series data. Inspired by the concept of utilizing DPO in fine-tuning unsupervised Language Models (LMs) to align with user preferences, `nanoDPO` pivots this approach to the realm of time series analysis. This library offers a unique perspective and toolset for time series forecasting, leveraging the principles of DPO to model and predict preferences in sequential data.

## Installation
To get started with `nanoDPO`, simply install the package using pip:

```bash
pip install nanoDPO
```

## Key Features

* Causal Transformer & Simple Sequence Model: Incorporates both a Causal Transformer and a LSTM-based Simple Sequence Model for diverse modeling needs.
* Preference Data Simulation: Utilizes a custom function, simulate_dpo_dataset_noise, to generate synthetic preference-based time series data.
* Sequence Data Preparation: Prepares data for training with prepare_sequence_datasets, aligning time series data with the DPO framework.
* DPO Training with PyTorch: Leverages the power of PyTorch for efficient and effective model training, complete with customizable parameters.
* MulticlassTrainer provides an additional approach to handle time series data, focusing on traditional multiclass classification tasks. 
* Cross-Entropy Loss for Multiclass Classification: Optimized for handling multiple classes in time series data.
* Customizable Training and Evaluation: Flexible parameters for epochs, batch size, and learning rate.
* Model Evaluation and Visualization: Offers tools for model evaluation and metrics visualization, ensuring an insightful analysis of performance.

## Usage

Import the necessary modules from nanoDPO, including the CausalTransformer, SimpleSequenceModel, and dataset preparation functions. Utilize the DPOOneModelTrainer for Direct Preference Optimization or MulticlassTrainer for conventional multiclass training.

```python
import torch
from nanodpo.causal_transformer import CausalTransformer
from nanodpo.simple_sequence_model import SimpleSequenceModel
from nanodpo.preference_data import simulate_dpo_dataset_noise
from nanodpo.sequence_data import prepare_sequence_datasets
from nanodpo.dpo_onemodel_trainer import DPOOneModelTrainer
from nanodpo.multiclass_trainer import MulticlassTrainer

# Initialize and train your model
...
model = CausalTransformer(d_feature= feature_dim, d_model=d_model, n_head=n_head, n_layer=n_layer, 
                          num_actions=num_actions, max_len=sequence_len,
                          device=device).to(device)
...
trainer = DPOOneModelTrainer(model=model, model_dir=f"dpo_{model_type}_model/", device=device,
                             learning_rate=learning_rate, batch_size=batch_size)
trainer.train(train_dataset, test_dataset, epochs=epochs, eval_interval=eval_interval)
# Evaluate and visualize the results
trainer.plot_metrics()
trainer.evaluate(test_dataset)
```

![wandb dpo causal_transformer](https://github.com/jamesliu/nanoDPO/blob/main/assets/dpo_causal_transformer_wandb.png)

## License

nanoDPO is licensed under the Apache License 2.0. See the [LICENSE](https://github.com/jamesliu/nanoDPO/blob/main/LICENSE) file for more details.

## Acknowledgments

Inspired by the paper "[Direct Preference Optimization: Your Language Model is Secretly a Reward Model](https://arxiv.org/abs/2305.18290)," nanoDPO extends the concept of DPO to the domain of time series data, opening new avenues for research and application.

## Citation

```bibtex
@misc{rafailov2023direct,
  title   = {Direct Preference Optimization: Your Language Model is Secretly a Reward Model}, 
  author  = {Rafael Rafailov and Archit Sharma and Eric Mitchell and Stefano Ermon and Christopher D. Manning and Chelsea Finn},
  year    = {2023},
  eprint  = {2305.18290},
  archivePrefix = {arXiv},
  primaryClass = {cs.LG}
}
```


docs/changelog.rst
.. _changelog:

===========
 Changelog
===========

.. _v0_1:

0.1 (2023-11-01)
----------------

- Initial release to PyPI
- Causal Transformer & Simple Sequence Model: Incorporates both a Causal Transformer and a LSTM-based Simple Sequence Model for diverse modeling needs.
- Preference Data Simulation: Utilizes a custom function, simulate_dpo_dataset_noise, to generate synthetic preference-based time series data.
- Sequence Data Preparation: Prepares data for training with prepare_sequence_datasets, aligning time series data with the DPO framework.
- DPO Training with PyTorch: Leverages the power of PyTorch for efficient and effective model training, complete with customizable parameters.
- MulticlassTrainer provides an additional approach to handle time series data, focusing on traditional multiclass classification tasks. 
- Cross-Entropy Loss for Multiclass Classification: Optimized for handling multiple classes in time series data.
- Customizable Training and Evaluation: Flexible parameters for epochs, batch size, and learning rate.
- Model Evaluation and Visualization: Offers tools for model evaluation and metrics visualization, ensuring an insightful analysis of performance.


tests/test_causal_transformer.py
import torch
from nanodpo.causal_transformer import CausalTransformer

def test_initialization():
    # Test initialization with different configurations
    model = CausalTransformer(d_feature=10, d_model=32, n_head=4, n_layer=2)
    assert model is not None, "Failed to initialize CausalTransformer"

def test_forward_pass():
    # Create a dummy input
    dummy_input = torch.rand(5, 10, 10)  # Batch size = 5, Sequence length = 10, Feature size = 10
    model = CausalTransformer(d_feature=10, d_model=32, n_head=4, n_layer=2)

    # Test forward pass
    try:
        output = model(dummy_input)
        assert output is not None, "Forward pass returned None"
    except Exception as e:
        assert False, f"Forward pass raised an exception: {e}"

def test_output_shape():
    # Define input parameters
    batch_size = 5
    seq_length = 10
    d_feature = 10
    num_actions = 3  # As defined in the CausalTransformer class

    # Create a dummy input
    dummy_input = torch.rand(batch_size, seq_length, d_feature)
    model = CausalTransformer(d_feature=d_feature, d_model=32, n_head=4, n_layer=2, num_actions=num_actions)

    # Test output shape
    output = model(dummy_input)
    expected_shape = (batch_size, num_actions)
    assert output.shape == expected_shape, f"Output shape is incorrect. Expected: {expected_shape}, Got: {output.shape}"



nanodpo/sequence_data.py
import torch
from sklearn.model_selection import train_test_split

# Function to prepare sequence datasets(TimeSeries data) 
def prepare_sequence_datasets(dataset, action_to_idx, sequence_len, step_size, test_size=0.2):
    # Adjust the loopback period according to step_size
    loopback_period = sequence_len * step_size

    # Convert dataset into features, winning (preferred) actions, and losing (rejected) actions
    features = [torch.tensor([data[0] for data in dataset[i:i+loopback_period:step_size]], dtype=torch.float32) 
                for i in range(len(dataset) - loopback_period + 1)]

    winning_actions = [torch.tensor(action_to_idx[dataset[i+loopback_period-step_size][1]], dtype=torch.long) 
                       for i in range(len(dataset) - loopback_period + 1)]

    losing_actions = [torch.tensor(action_to_idx[dataset[i+loopback_period-step_size][2]], dtype=torch.long) 
                      for i in range(len(dataset) - loopback_period +1)]

    # Combine winning and losing actions for splitting
    actions = list(zip(winning_actions, losing_actions))

    # Splitting the dataset
    X_train, X_test, actions_train, actions_test = train_test_split(features, actions, test_size=test_size)

    # Separate winning and losing actions after splitting
    yw_train, yl_train = zip(*actions_train)
    yw_test, yl_test = zip(*actions_test)

    # Create PyTorch datasets
    train_dataset = list(zip(X_train, yw_train, yl_train))
    test_dataset = list(zip(X_test, yw_test, yl_test))

    return train_dataset, test_dataset


nanodpo/preference_data.py
# filename: preference_data.py

import numpy as np
import random

# Preference data generation Yw > Yl
def simulate_dpo_dataset_simple(size=1000, feature_dim=3):
    dataset = []
    actions = ['POSITIVE', 'NEGATIVE', 'NEUTRAL']

    for _ in range(size):
        # Randomly choose an action
        Yw = random.choice(actions)

        # Generate features based on the action
        if Yw == 'POSITIVE':
            # Example pattern for POSITIVE
            features = np.random.normal(0.5, 0.05, feature_dim)
        elif Yw == 'NEGATIVE':
            # Example pattern for NEGATIVE
            features = np.random.normal(-0.5, 0.05, feature_dim)
        elif Yw == 'NEUTRAL':
            # Example pattern for NEUTRAL 
            features = np.random.normal(0.0, 0.05, feature_dim)
        else:
            raise ValueError(f"Unknown action: {Yw}")

        for Yl in actions:
            if Yl != Yw:
                # Add to dataset
                dataset.append((features, Yw, Yl))

    return dataset, actions

def simulate_dpo_dataset_noise(size=1000, feature_dim=3):
    dataset = []
    actions = ['POSITIVE', 'NEGATIVE', 'NEUTRAL']

    for _ in range(size):
        Yw = random.choice(actions)

        if Yw == 'POSITIVE':
            base = np.random.normal(-0.9, 1, feature_dim)
            noise = np.random.normal(0, 0.05, feature_dim)
            transformation = np.sin(np.linspace(-np.pi, np.pi, feature_dim))
            features = base + noise + transformation

        elif Yw == 'NEGATIVE':
            base = np.random.normal(-0.5, 1, feature_dim)
            noise = np.random.normal(0, 0.05, feature_dim)
            transformation = np.cos(np.linspace(-np.pi - 0.5 , np.pi + 0.5, feature_dim))
            features = base + noise + transformation

        elif Yw == 'NEUTRAL':
            base = np.random.uniform(-0.1, 1, feature_dim)
            noise = np.random.normal(0, 0.05, feature_dim)
            transformation = np.tanh(np.linspace(-1, 1, feature_dim))
            features = base + noise + transformation

        else:
            raise ValueError(f"Unknown action: {Yw}")

        for Yl in actions:
            if Yl != Yw:
                dataset.append((features, Yw, Yl))

    return dataset, actions


if __name__ == '__main__':

    dataset, actions = simulate_dpo_dataset()
    print('dataset', len(dataset))
    for i, data in enumerate(dataset):
        print(data)
        if i > 5:
            break
    

nanodpo/causal_transformer.py
# filename: causal_transformer.py

import torch
import torch.nn as nn
import torch.nn.functional as F
from nanodpo.sinusoidal_positional_encoding import SinusoidalPositionalEncoding

class CausalTransformerLayer(nn.Module):
    def __init__(self, d_model, n_head, dim_feedforward=1024, dropout=0.1):
        super(CausalTransformerLayer, self).__init__()
        self.self_attn = nn.MultiheadAttention(d_model, n_head, dropout=dropout)

        # Implementation of Feedforward model
        self.linear1 = nn.Linear(d_model, dim_feedforward)
        self.dropout = nn.Dropout(dropout)
        self.linear2 = nn.Linear(dim_feedforward, d_model)

        self.norm1 = nn.LayerNorm(d_model)
        self.norm2 = nn.LayerNorm(d_model)
        self.dropout1 = nn.Dropout(dropout)
        self.dropout2 = nn.Dropout(dropout)

    def forward(self, src, src_mask=None, src_key_padding_mask=None):
        src2 = self.norm1(src)
        q = k = src2
        src2 = self.self_attn(q, k, value=src2, attn_mask=src_mask,
                              key_padding_mask=src_key_padding_mask)[0]
        src = src + self.dropout1(src2)
        src2 = self.norm2(src)
        src2 = self.linear2(self.dropout(F.relu(self.linear1(src2))))
        src = src + self.dropout2(src2)
        return src

class CausalTransformer(nn.Module):
    def __init__(self, d_feature, d_model, n_head, n_layer, num_actions=3, dim_feedforward=2048, dropout=0.1, 
                 max_len=5000, device=torch.device("cpu")):
        super(CausalTransformer, self).__init__()
        self.device = device  # Store the device
        self.embedding = nn.Linear(d_feature, d_model)
        self.layers = nn.ModuleList([CausalTransformerLayer(d_model, n_head, dim_feedforward, dropout) 
                                     for _ in range(n_layer)])
        self.positional_encoding = SinusoidalPositionalEncoding(d_model, max_len, device=self.device).to(self.device)
        self.output_layer = nn.Linear(d_model, num_actions)  # Output layer for three actions

    def forward(self, src, src_mask=None, src_key_padding_mask=None):
        src = self.embedding(src)
        src = self.positional_encoding(src)
        for layer in self.layers:
            src = layer(src, src_mask, src_key_padding_mask)

        output = self.output_layer(src)
        # Optional: Apply softmax for probabilities
        # output = F.softmax(output, dim=-1)
        output = output[:, -1, :]  # Shape: [batch_size, num_actions]
        return output


nanodpo/multiclass_trainer.py
import os
import torch
from torch.utils.data import DataLoader
import torch.optim as optim
from torch.optim.lr_scheduler import StepLR
import matplotlib.pyplot as plt
import wandb
from nanodpo.causal_transformer import CausalTransformer
from nanodpo.simple_sequence_model import SimpleSequenceModel
from nanodpo.preference_data import simulate_dpo_dataset_noise
from nanodpo.sequence_data import prepare_sequence_datasets
import warnings
warnings.filterwarnings("ignore")

class MulticlassTrainer:
    def __init__(self, model, model_dir, device=torch.device("cpu"), learning_rate=0.001, batch_size=32):
        self.model = model
        self.model_dir = model_dir
        self.device = device
        self.criterion = torch.nn.CrossEntropyLoss().to(self.device)
        self.optimizer = optim.Adam(self.model.parameters(), lr=learning_rate)
        self.batch_size = batch_size
        self.loss_history = []
        self.accuracy_history = []
        self.model_dir = model_dir
        os.makedirs(self.model_dir, exist_ok=True)

    def train(self, train_dataset, test_dataset, epochs=10, eval_interval=2, step_size=10, gamma=0.1):
        train_loader = DataLoader(train_dataset, batch_size=self.batch_size, shuffle=True)
        self.model.train()
        # Initialize the scheduler
        scheduler = StepLR(self.optimizer, step_size=step_size, gamma=gamma)
        
        for epoch in range(epochs):
            total_loss = 0
            for i, (inputs, yw, yl) in enumerate(train_loader):
                # Data preparation steps
                inputs = inputs.to(self.device)
                yw = yw.to(self.device)
                yl = yl.to(self.device)

                # Reset gradients
                self.optimizer.zero_grad()

                # Forward pass
                outputs = self.model(inputs)
            
                # Compute loss
                loss = self.criterion(outputs, yw)
                total_loss += loss.item()

                # Backward pass and optimize
                loss.backward()
                self.optimizer.step()

                # Diagnostic information
                if ((i+1) % 1000 == 0) and ((epoch + 1) % eval_interval == 0):
                    self.log_diagnostics(inputs, yw, outputs)
                    self.print_predictions(outputs, yw)

            # Step the scheduler after each epoch
            #scheduler.step()

            current_lr = scheduler.get_last_lr()[0]

            epoch_loss = total_loss / len(train_loader)
            self.loss_history.append(epoch_loss)


            # Evaluate at specified intervals
            if (epoch + 1) % eval_interval == 0 or epoch == epochs - 1:
                print(f"Current Learning Rate: {current_lr}")
                print(f'Epoch [{epoch+1}/{epochs}], Loss: {epoch_loss:.4f}')
                # Evaluate the model
                self.model.eval()
                accuracy = self.evaluate(test_dataset)
                self.model.train()  # Switch back to training mode
                self.accuracy_history.append(accuracy)
                # Save model after evaluation
                self.save_model(epoch + 1)
                # Log metrics to wandb
                wandb.log({"epoch": epoch, "loss": epoch_loss, "accuracy": accuracy})

    def evaluate(self, test_dataset):
        test_loader = DataLoader(test_dataset, batch_size=self.batch_size, shuffle=False)
        self.model.eval()

        total_correct = 0
        total_samples = 0
        action_count = {'preferred': {}, 'rejected': {}, 'predicted': {}}

        with torch.no_grad():
            for inputs, preferred_actions, rejected_actions in test_loader:
                inputs = inputs.to(self.device)
                preferred_actions = preferred_actions.to(self.device)
                rejected_actions = rejected_actions.to(self.device)

                outputs = self.model(inputs)

                _, predicted = torch.max(outputs.data, 1)
                total_samples += preferred_actions.size(0)
                total_correct += (predicted == preferred_actions).sum().item()

                for action in torch.unique(preferred_actions).tolist():
                    action_count['preferred'][action] = action_count['preferred'].get(action, 0) + (preferred_actions == action).sum().item()
    
                for action in torch.unique(rejected_actions).tolist():
                    action_count['rejected'][action] = action_count['rejected'].get(action, 0) + (rejected_actions == action).sum().item()
    
                for action in torch.unique(predicted).tolist():
                    action_count['predicted'][action] = action_count['predicted'].get(action, 0) + (predicted == action).sum().item()
    
        accuracy = (total_correct / total_samples) * 100
        print(f'Accuracy: {accuracy}%')

        # Diagnostic information
        print("Action Counts:")
        for action_type, counts in action_count.items():
            print(f"{action_type.capitalize()}: {counts}")

        return accuracy
    
    def plot_metrics(self, eval_interval=2):
        plt.figure(figsize=(12, 5))

        # Plotting Loss
        plt.subplot(1, 2, 1)
        plt.plot(self.loss_history, label='Loss')
        plt.xlabel('Epochs')
        plt.ylabel('Loss')
        plt.title('Training Loss')
        plt.legend()

        # Adjusting x-axis for Accuracy
        accuracy_x_axis = list(range(eval_interval, epochs + 1, eval_interval))
        if epochs % eval_interval != 0:
            accuracy_x_axis.append(epochs)  # Add the last epoch if it doesn't align with the interval
    
        # Plotting Accuracy
        plt.subplot(1, 2, 2)
        plt.plot(accuracy_x_axis, self.accuracy_history, label='Accuracy')
        plt.xlabel('Epochs')
        plt.ylabel('Accuracy (%)')
        plt.title('Evaluation Accuracy')
        plt.legend()
        
        plt.show()
        plt.savefig('loss_accuracy.png')

    def save_model(self, epoch):
        torch.save(self.model.state_dict(), os.path.join(self.model_dir, f"model_epoch_{epoch}.pt"))

    def load_model(self, model_version=None):
        if model_version is not None:
            model_path = os.path.join(self.model_dir, f"model_epoch_{model_version}.pt")
        else:
            model_files = [f for f in os.listdir(self.model_dir) if f.startswith("model_epoch_") and f.endswith(".pt")]
            if not model_files:
                return  # No model to load
            latest_model = max(model_files, key=lambda x: int(x.split('_')[2].split('.')[0]))
            model_path = os.path.join(self.model_dir, latest_model)

        if os.path.exists(model_path):
            self.model.load_state_dict(torch.load(model_path, map_location=self.device))

    def log_diagnostics(self, inputs, labels, outputs):
        # Gradient norms
        total_norm = 0
        for p in self.model.parameters():
            param_norm = p.grad.data.norm(2)
            total_norm += param_norm.item() ** 2
        total_norm = total_norm ** 0.5

        print(f"Gradient Norm: {total_norm}")

        # Output distribution
        print(f"Output distribution: {outputs.data}")

        # Input and Label Statistics
        print(f"Input stats: Mean={inputs.mean()}, Std={inputs.std()}")
        print(f"Label distribution: {torch.bincount(labels)}")

    def print_weights(self):
        for name, param in self.model.named_parameters():
            if param.requires_grad:
                print(f"{name}: {param.data}")
    
    def print_predictions(self, outputs, labels):
        _, predicted = torch.max(outputs.data, 1)
        print(f"Predicted: {predicted.tolist()}, Actual: {labels.tolist()}")
    
    def print_model_summary(self):
        print("Model Summary:")
        print(self.model)
        print("\n\nNumber of parameters: ", sum(p.numel() for p in self.model.parameters()))

if __name__ == '__main__':
    config = {'learning_rate':0.001, 'batch_size': 32, 'sequence_len': 10, 'step_size': 5}
    size = 1000
    feature_dim = 3 
    dataset, actions = simulate_dpo_dataset_noise(size=size, feature_dim=feature_dim)

    num_actions = len(actions)
    feature = dataset[0][0]
    sequence_len = config['sequence_len'] 
    step_size = config['step_size']
    epochs = 100
    eval_interval = 10
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model_type = 'transformer'
    if model_type == 'transformer':
        n_head = 2
        d_model = 16
        n_layer =2
        assert(d_model % n_head == 0)
        model = CausalTransformer(d_feature=feature_dim, d_model=d_model, n_head=n_head, n_layer=n_layer, 
                                       num_actions=num_actions, max_len=sequence_len,
                                       device=device).to(device)
    elif model_type == 'lstm':
        hidden_dim = 64
        model = SimpleSequenceModel(feature_dim=feature_dim, hidden_dim=hidden_dim, 
                                    num_actions=num_actions).to(device)
    else:
        raise ValueError(f"Unknown model type: {model_type}")
    trainer = MulticlassTrainer(model=model, model_dir=f"multiclass_{model_type}_model/", device=device,
                                 learning_rate=config["learning_rate"], batch_size=config["batch_size"])
    # Assuming 'dataset' is your data in the required format
    action_to_idx = {action:i for i,action in enumerate(actions)}
    train_dataset, test_dataset = prepare_sequence_datasets(dataset, action_to_idx=action_to_idx, 
                                                            sequence_len=sequence_len ,step_size=step_size)
    
    # Initialize Weights & Biases
    wandb.init(project='nanoDPO', name=f"multiclass_{model_type}")
                                                        
    # Now you can use train_dataset and test_dataset for training and evaluation
    trainer.train(train_dataset, test_dataset, epochs=epochs, eval_interval=eval_interval)
    trainer.evaluate(test_dataset)

    wandb.finish()


nanodpo/simple_sequence_model.py
import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np

class SimpleSequenceModel(nn.Module):
    def __init__(self, feature_dim, hidden_dim, num_actions):
        super(SimpleSequenceModel, self).__init__()
        self.lstm = nn.LSTM(input_size=features_dim, hidden_size=hidden_dim, batch_first=True)
        self.fc = nn.Linear(hidden_dim, num_actions)

    def forward(self, x):
        # x shape: (batch, seq_len, features_dim)
        lstm_out, _ = self.lstm(x)
        # lstm_out shape: (batch, seq_len, hidden_dim)
        
        # We take the output of the last time step
        last_time_step = lstm_out[:, -1, :]
        # last_time_step shape: (batch, hidden_dim)
        
        out = self.fc(last_time_step)
        # out shape: (batch, num_actions)
        return out

def generate_simple_dataset(size=500, seq_len=10, num_features=1):
    data = np.zeros((size, seq_len, num_features))
    labels = np.zeros(size, dtype=int)  # Assuming three classes: 0 (BUY), 1 (SELL), 2 (HOLD)

    for i in range(size):
        pattern_type = np.random.choice(['increasing', 'decreasing', 'constant'])
        if pattern_type == 'increasing':
            data[i, :, 0] = np.linspace(0, 1, seq_len)
            labels[i] = 0  # 'BUY'
        elif pattern_type == 'decreasing':
            data[i, :, 0] = np.linspace(1, 0, seq_len)
            labels[i] = 1  # 'SELL'
        else:  # 'constant'
            constant_value = np.random.uniform(0, 1)
            data[i, :, 0] = constant_value
            labels[i] = 2  # 'HOLD'

    data_tensor = torch.tensor(data, dtype=torch.float32)
    labels_tensor = torch.tensor(labels, dtype=torch.long)

    return data_tensor, labels_tensor

# Function to train the model
def train_model(model, data, labels, epochs=50, batch_size=32, learning_rate=0.001):
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=learning_rate)
    
    for epoch in range(epochs):
        total_loss = 0
        model.train()

        for i in range(0, len(data), batch_size):
            inputs = data[i:i + batch_size]
            target = labels[i:i + batch_size]

            optimizer.zero_grad()
            outputs = model(inputs)
            loss = criterion(outputs, target)
            loss.backward()
            optimizer.step()

            total_loss += loss.item()

        print(f'Epoch {epoch+1}/{epochs}, Loss: {total_loss/len(data)}')

if __name__ == '__main__':
    features_dim = 3
    seq_len = 10
    data_tensor, labels_tensor = generate_simple_dataset(seq_len=seq_len, num_features=features_dim)

    # Define the model
    hidden_dim = 64
    num_actions = 3
    model = SimpleSequenceModel(features_dim, hidden_dim, num_actions)

    # Train the model
    train_model(model, data_tensor, labels_tensor)

nanodpo/dpo_onemodel_trainer.py
import os
import torch
import torch.nn as nn
import torch.optim as optim
from torch.optim.lr_scheduler import StepLR
from torch.utils.data import DataLoader
import matplotlib.pyplot as plt
import wandb
from nanodpo.causal_transformer import CausalTransformer
from nanodpo.simple_sequence_model import SimpleSequenceModel
from nanodpo.preference_data import simulate_dpo_dataset_noise
from nanodpo.sequence_data import prepare_sequence_datasets
import warnings
warnings.filterwarnings("ignore")

class DPOOneModelTrainer:
    def __init__(self, model, model_dir, device=torch.device("cpu"), learning_rate=0.005, batch_size=32, margin=1.0):
        self.model = model
        self.device = device
        self.optimizer = optim.Adam(self.model.parameters(), lr=learning_rate)
        self.batch_size = batch_size
        # Model, optimizer setup remains the same
        self.criterion = nn.MarginRankingLoss(margin=margin).to(self.device)
        self.loss_history = []
        self.accuracy_history = []
        self.eval_interval = -1
        self.model_dir = model_dir
        os.makedirs(self.model_dir, exist_ok=True)

    def save_model(self, epoch):
        torch.save(self.model.state_dict(), os.path.join(self.model_dir, f"model_epoch_{epoch}.pt"))

    def load_model(self, model_version=None):
        if model_version is not None:
            model_path = os.path.join(self.model_dir, f"model_epoch_{model_version}.pt")
        else:
            model_files = [f for f in os.listdir(self.model_dir) if f.startswith("model_epoch_") and f.endswith(".pt")]
            if not model_files:
                return  # No model to load
            latest_model = max(model_files, key=lambda x: int(x.split('_')[2].split('.')[0]))
            model_path = os.path.join(self.model_dir, latest_model)

        if os.path.exists(model_path):
            self.model.load_state_dict(torch.load(model_path, map_location=self.device))

    def train(self, train_dataset, test_dataset, epochs=10, eval_interval=2, step_size=10, gamma=0.1):
        self.eval_interval = eval_interval
        train_loader = DataLoader(train_dataset, batch_size=self.batch_size, shuffle=True)
        # DataLoader setup remains the same
        self.model.train()

        # Initialize the scheduler
        scheduler = StepLR(self.optimizer, step_size=step_size, gamma=gamma)

        for epoch in range(epochs):
            total_loss = 0
            for inputs, yw, yl in train_loader:
                inputs = inputs.to(self.device)
                yw = yw.to(self.device)
                yl = yl.to(self.device)

                # Data prep steps
                self.optimizer.zero_grad()

                # Forward pass
                outputs = self.model(inputs)
                # Get scores for preferred and rejected actions
                preferred_scores = outputs[torch.arange(outputs.size(0)), yw]
                rejected_scores = outputs[torch.arange(outputs.size(0)), yl]

                # Targets for MarginRankingLoss (all +1s, as preferred should be ranked higher)
                targets = torch.ones(preferred_scores.size(0), device=preferred_scores.device)

                # Compute loss
                loss = self.criterion(preferred_scores, rejected_scores, targets)
                total_loss += loss.item()

                # Backward pass and optimize
                loss.backward()
                self.optimizer.step()
            
            current_lr = scheduler.get_last_lr()[0]

            epoch_loss = total_loss / len(train_loader)
            self.loss_history.append(epoch_loss)

            # Evaluate at specified intervals
            if (epoch + 1) % eval_interval == 0 or epoch == epochs - 1:
                print(f"Current Learning Rate: {current_lr}")
                print(f'Epoch [{epoch+1}/{epochs}], Loss: {epoch_loss}')
                # Evaluate the model
                self.model.eval()
                accuracy = self.evaluate(test_dataset)
                self.model.train()  # Switch back to training mode
                self.accuracy_history.append(accuracy)
                # Save model after evaluation
                self.save_model(epoch + 1)

                # Log metrics to wandb
                wandb.log({"epoch": epoch, "loss": epoch_loss, "accuracy": accuracy})

    # Evaluation function can remain the same or be adapted for ranking-based metrics
    def evaluate(self, test_dataset):
        test_loader = DataLoader(test_dataset, batch_size=self.batch_size, shuffle=False)
        self.model.eval()

        total_correct = 0
        total_samples = 0
        action_count = {'preferred': {}, 'rejected': {}, 'predicted': {}}

        with torch.no_grad():
            for inputs, preferred_actions, rejected_actions in test_loader:
                inputs = inputs.to(self.device)
                preferred_actions = preferred_actions.to(self.device)
                rejected_actions = rejected_actions.to(self.device)
                outputs = self.model(inputs)

                _, predicted = torch.max(outputs.data, 1)
                for action in torch.unique(preferred_actions).tolist():
                    action_count['preferred'][action] = action_count['preferred'].get(action, 0) + (preferred_actions == action).sum().item()
    
                for action in torch.unique(rejected_actions).tolist():
                    action_count['rejected'][action] = action_count['rejected'].get(action, 0) + (rejected_actions == action).sum().item()
    
                for action in torch.unique(predicted).tolist():
                    action_count['predicted'][action] = action_count['predicted'].get(action, 0) + (predicted == action).sum().item()
    
                total_samples += preferred_actions.size(0)
                total_correct += (predicted == preferred_actions).sum().item()

        accuracy = (total_correct / total_samples) * 100
        print(f'Accuracy: {accuracy}%')

        # Diagnostic information
        print("Action Counts:")
        for action_type, counts in action_count.items():
            print(f"{action_type.capitalize()}: {counts}")
    
        return accuracy
    
    def plot_metrics(self):
        if self.eval_interval < 1:
            return
        plt.figure(figsize=(12, 5))

        # Plotting Loss
        plt.subplot(1, 2, 1)
        plt.plot(self.loss_history, label='Loss')
        plt.xlabel('Epochs')
        plt.ylabel('Loss')
        plt.title('Training Loss')
        plt.legend()

        # Adjusting x-axis for Accuracy
        accuracy_x_axis = list(range(self.eval_interval, epochs + 1, self.eval_interval))
        if epochs % eval_interval != 0:
            accuracy_x_axis.append(epochs)  # Add the last epoch if it doesn't align with the interval
    
        # Plotting Accuracy
        plt.subplot(1, 2, 2)
        plt.plot(accuracy_x_axis, self.accuracy_history, label='Accuracy')
        plt.xlabel('Epochs')
        plt.ylabel('Accuracy (%)')
        plt.title('Evaluation Accuracy')
        plt.legend()
        
        plt.show()
        plt.savefig('loss_accuracy.png')

if __name__ == '__main__':
    config = {'learning_rate':0.001, 'batch_size': 32, 'sequence_len': 10, 'step_size': 5}
    size = 1000
    feature_dim = 3
    dataset, actions = simulate_dpo_dataset_noise(size=size, feature_dim=feature_dim)
    print(f"Dataset size: {len(dataset)}")
    num_actions = len(actions)
    feature = dataset[0][0]
    learning_rate = config['learning_rate']
    batch_size = config['batch_size']
    sequence_len = config['sequence_len'] 
    step_size = config['step_size']
    epochs = 100
    eval_interval = 1
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model_type = 'transformer'
    if model_type == 'transformer':
        n_head = 2
        d_model = 4
        n_layer = 2
        assert(d_model % n_head == 0)
        model = CausalTransformer(d_feature= feature_dim, d_model=d_model, n_head=n_head, n_layer=n_layer, 
                                       num_actions=num_actions, max_len=sequence_len,
                                       device=device).to(device)
    elif model_type == 'lstm':
        hidden_dim = 64
        model = SimpleSequenceModel(feature_dim=feature_dim, hidden_dim=hidden_dim, 
                                    num_actions=num_actions).to(device)
    else:
        raise ValueError(f"Unknown model type: {model_type}")
    
    # Assuming 'dataset' is your data in the required format
    action_to_idx = {action:i for i,action in enumerate(actions)}
    train_dataset, test_dataset = prepare_sequence_datasets(dataset, action_to_idx=action_to_idx,
                                                            sequence_len=sequence_len, step_size=step_size)

    model_version = None  # Set to a specific version if needed
    trainer = DPOOneModelTrainer(model=model, model_dir=f"dpo_{model_type}_model/", device=device,
                                 learning_rate=learning_rate, batch_size=batch_size)
    trainer.load_model(model_version)

    # Initialize Weights & Biases
    wandb.init(project='nanoDPO', name=f"dpo_{model_type}")

    # Now you can use train_dataset and test_dataset for training and evaluation
    trainer.train(train_dataset, test_dataset, epochs=epochs, eval_interval=eval_interval)
    trainer.plot_metrics(eval_interval)
    trainer.evaluate(test_dataset)

    wandb.finish()